Text/CustomEntitySearch/CustomEntitySearch.cs

```diff
@@ -52,48 +53,273 @@ public static class CustomEntitySearch
+                     {
+                         outRecord.Errors.Add(new WebApiErrorWarningContract { Message = "Cannot process record without the given key 'words' in the dictionary"});
+                     }
+                     else
+                     {
+                         string text = inRecord.Data["text"] as string;
+                         IList<string> words = (inRecord.Data.ContainsKey("words")) ? 
+                         ((JArray)inRecord.Data["words"]).ToObject<List<string>>() : new List<string>();
+                         Dictionary<string, string[]> synonyms = (inRecord.Data.ContainsKey("synonyms")) ? 
+                         ((JContainer)inRecord.Data["synonyms"]).ToObject<Dictionary<string, string[]>>() : new Dictionary<string, string[]>();
+                         IList<string> exactMatches = (inRecord.Data.ContainsKey("exactMatches")) ? 
+                         ((JArray)inRecord.Data["exactMatches"]).ToObject<List<string>>() : new List<string>();
+                         int offset = (inRecord.Data.ContainsKey("fuzzyMatchOffset") && (int)(long)inRecord.Data["fuzzyMatchOffset"] >= 0) ? 
+                         (int)(long)inRecord.Data["fuzzyMatchOffset"] : 0;
+                         bool caseSensitive = (inRecord.Data.ContainsKey("caseSensitive")) ? (bool)inRecord.Data.ContainsKey("caseSensitive") : false;
+                         if (words.Count == 0 || (words.Count == 1 && words[0] == ""))
+                         {
+                             outRecord.Warnings.Add(new WebApiErrorWarningContract { Message = "The given key 'words' was not present in the dictionary." });
+                             WordLinker userInput = new WordLinker(executionContext.FunctionAppDirectory);
+                             words = userInput.Words;
+                             synonyms = userInput.Synonyms;
+                             exactMatches = userInput.ExactMatch;
+                             offset = (userInput.FuzzyMatchOffset >= 0) ? userInput.FuzzyMatchOffset : 0;
+                             caseSensitive = userInput.CaseSensitive;
+                         }
+
+                         var entities = new List<Entity>();
+                         var entitiesFound = new HashSet<string>();
+                         if (!string.IsNullOrWhiteSpace(text))
+                         {
+                             foreach (string word in words)
+                             {
+                                 if (string.IsNullOrEmpty(word)) continue;
+                                 int leniency = (exactMatches != null && exactMatches.Contains(word)) ? 0 : offset;
+                                 AddValues(word, text, entities, entitiesFound, leniency, caseSensitive);
+                                 if (synonyms != null && synonyms.ContainsKey(word))
```

Comment:
You can improve this a bit with `synonyms.TryGetValue(word, out string[] wordSynonyms)`.